src/Strategy/Dart/PubSpec.hs

```diff
@@ -0,0 +1,178 @@
+  parseJSON = Yaml.withObject "pubspec.yaml content" $ \o -> do
+    dependencies <- o .:? "dependencies"
+    devDependencies <- o .:? "dev_dependencies"
+    depOverrides <- o .:? "dependency_overrides"
+    pure $ PubSpecContent dependencies devDependencies depOverrides
+
+-- Formatter crashes here.
+{- ORMOLU_DISABLE -}
+instance FromJSON PubSpecDepSource where
+  parseJSON (Yaml.String s) = pure $ PubSpecDepHostedSource (Just . CEq $ s) Nothing Nothing
+  parseJSON (Yaml.Object o) =
+    asum
+      [ PubSpecDepHostedSource . Just . CEq
+          <$> o .: "version"
+          <*> o .: "hosted" |> "name"
+          <*> o .: "hosted" |> "url"
+      , PubSpecDepGitSource . Just . CEq
+          <$> o .: "git" |> "ref"
+          <*> o .: "git" |> "url"
+      , PubSpecDepGitSource Nothing
+          <$> o .: "git"
+      , PubSpecDepPathSource
+          <$> o .: "path"
+      , PubSpecDepSdkSource
+          <$> o .: "sdk"
+      ]
+    where
+      (|>) :: FromJSON a => Yaml.Parser Yaml.Object -> Text -> Yaml.Parser a
+      (|>) parser key = do
+        obj <- parser
+        obj .: key
+  parseJSON _ = fail "failed parsing pub package's source!"
+{- ORMOLU_ENABLE -}
+
+toDependency :: [DepEnvironment] -> PackageName -> PubSpecDepSource -> Dependency
+toDependency environments name (PubSpecDepHostedSource version _ url) =
+  Dependency
+    { dependencyType = PubType
+    , dependencyName = unPackageName name
+    , dependencyVersion = version
+    , dependencyLocations = maybeToList url
+    , dependencyEnvironments = environments
+    , dependencyTags = Map.empty
+    }
+toDependency environments _ (PubSpecDepGitSource gitRef gitUrl) =
+  Dependency
+    { dependencyType = GitType
+    , dependencyName = gitUrl
+    , dependencyVersion = gitRef
+    , dependencyLocations = []
+    , dependencyEnvironments = environments
+    , dependencyTags = Map.empty
+    }
+toDependency _ _ (PubSpecDepSdkSource _) = error "unexpected sdk dependency conversion"
+toDependency _ _ (PubSpecDepPathSource _) = error "unexpected path dependency conversion"
+
+isSupported :: PubSpecDepSource -> Bool
+isSupported PubSpecDepHostedSource{} = True
+isSupported (PubSpecDepGitSource _ _) = True
+isSupported _ = False
+
+buildGraph :: PubSpecContent -> Graphing.Graphing Dependency
+buildGraph specContent = foldr Graphing.direct Graphing.empty allDependencies
+  where
+    dependencies :: Map PackageName PubSpecDepSource
+    dependencies = fromMaybe Map.empty $ pubSpecDependencies specContent
+
+    devDependencies :: Map PackageName PubSpecDepSource
+    devDependencies = fromMaybe Map.empty $ pubSpecDevDependencies specContent
+
+    -- In pub manifest, dependency can be overriden.
+    -- Ref: https://dart.dev/tools/pub/dependencies#dependency-overrides
+    supersededDependencies :: Map PackageName PubSpecDepSource
+    supersededDependencies = fromMaybe Map.empty $ pubSpecDependenciesOverrides specContent
+
+    notSuperseded :: PackageName -> a -> Bool
```

Comment:
Along the same lines as my other comment wrt incremental transformation, I think we can use the same principles to untangle the "superseded" and "supported" logic here. Specifically, something like:

1. build a `Graphing PackageName` with all of the edges/etc
2. `gmap` in a function that turns them into `PubSpecDepSource`(?)
3. use `Graphing.shrink` to filter out unsupported packages
4. `gmap` in `toDependency`